### DISCUSSION TITLE
Applicative combinators instead of `construct!`

### DISCUSSION THREAD

**Discussion by mausch** [pacak/bpaf#333] [Q&A] [answered]
Hi and thanks for this library!
I'm relatively new to Rust and being a fan of optparse-applicative I was happy to find bpaf.
However I'm having issues with the `constructor!` macro, it's a bit too magicky for me as a Rust newbie...
I can't seem to find any combinators to build the `ParseCon` (I think?), is this possible at all without using the macro?
I know that currying isn't idiomatic in Rust but maybe bpaf already has some sort of lift2 / apply functions somewhere?

**Comment by pacak** [comment, marked as answer]
`construct!` macro works in two different modes: alternative and applicative combinations. For applicative you put parsers for all the variants into a set of square brackets and `construct!` combines them giving preference to one that consumes earlier item from the command line or one that succeds first in a list if they consume the same item. For applicative the easiest way to think about it is that instead of parsers (say `a: impl Parser<A>` and `b: impl Parser<B>`) you have variables (`a: A` and `b: B`) and you simply construct the result: `Foo { a, b }`. `construct!(Foo { a, b })` does the same but with Parsers, giving you `impl Parser<Foo>` back. A tiny bit of magic allows you to use `()` to specify that `a` is not a `Parser` by itself, but instead a function that produces one. As a beginner you might want to avoid using that bit of magic.

Rust doesn't play nicely with currying unfortunately: no higher order types, no curried type applications, lots of things are not `Copy`. I had some combinators exposed in earlier versions but decided against it since it was hard to use.
